src/newgrf_storage.h

```diff
@@ -104,10 +88,9 @@ struct PersistentStorageArray : BasePersistentStorageArray {
 
 		/* We do not have made a backup; lets do so */
 		if (AreChangesPersistent()) {
-			assert(this->prev_storage == nullptr);
-		} else if (this->prev_storage == nullptr) {
-			this->prev_storage = MallocT<TYPE>(SIZE);
-			memcpy(this->prev_storage, this->storage, sizeof(this->storage));
+			assert(!this->prev_storage);
+		} else if (!this->prev_storage) {
+			this->prev_storage = std::make_unique<StorageType>(this->storage);
```

Comment:
Why exactly a unique-pointer? If I read this correct, you kinda use it as an `std::optional`? Ofc you can use an unique-ptr as an optional, but initial I was thinking you actually need the pointer. But you just want to know if the object is set yes/no?

Reply:
As far as I know, using std::optional<> would (may?) mean prev_storage always takes up space even if it's not used. Given this 1KB each, using std::unique_ptr<> instead made sense to me.